google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/gaxx/grpc/ChannelPoolHealthChecker.java

```diff
@@ -0,0 +1,229 @@
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/** Class that manages the health checking in the BigtableChannelPool */
+public class ChannelPoolHealthChecker {
+
+  // Configuration constants
+  // Window_Duration is the duration over which we keep probe results
+  private static final Duration WINDOW_DURATION = Duration.ofMinutes(5);
+  // Interval at which we probe channel health
+  static final Duration PROBE_INTERVAL = Duration.ofSeconds(30);
+  // Timeout deadline for a probe
+  @VisibleForTesting static final Duration PROBE_DEADLINE = Duration.ofMillis(500);
+  // Minimum interval between new idle channel evictions
+  private static final Duration MIN_EVICTION_INTERVAL = Duration.ofMinutes(10);
+  // Minimum number of probes that must be sent to a channel before it will be considered for
+  // eviction
+  private static final int MIN_PROBES_FOR_EVALUATION = 4;
+  // Percentage of probes that must fail for a channel to be considered unhealthy
+  private static final int SINGLE_CHANNEL_FAILURE_PERCENT_THRESHOLD = 60;
+  // "Circuitbreaker" - If this or a higher percentage of channels in a pool are bad, we will not
+  // evict any channels
+  private static final int POOLWIDE_BAD_CHANNEL_CIRCUITBREAKER_PERCENT = 70;
+
+  /** Inner class to represent the result of a single probe. */
+  @AutoValue
+  abstract static class ProbeResult {
+    abstract Instant startTime();
+
+    abstract boolean isSuccessful();
+
+    static ProbeResult create(Instant startTime, boolean success) {
+      return new AutoValue_ChannelPoolHealthChecker_ProbeResult(startTime, success);
+    }
+  }
+
+  private final Supplier<ImmutableList<Entry>> entrySupplier;
+  private volatile Instant lastEviction;
+  private ScheduledExecutorService executor;
+
+  private ChannelPrimer channelPrimer;
+
+  private final Clock clock;
+
+  /** Constructor for the pool health checker. */
+  public ChannelPoolHealthChecker(
+      Supplier<ImmutableList<Entry>> entrySupplier,
+      ChannelPrimer channelPrimer,
+      ScheduledExecutorService executor,
+      Clock clock) {
+    this.entrySupplier = entrySupplier;
+    this.lastEviction = Instant.MIN;
+    this.channelPrimer = channelPrimer;
+    this.executor = executor;
+    this.clock = clock;
+  }
+
+  void start() {
+    Duration initialDelayProbe =
+        Duration.ofMillis(ThreadLocalRandom.current().nextLong(PROBE_INTERVAL.toMillis()));
+    executor.scheduleAtFixedRate(
+        this::runProbes,
+        initialDelayProbe.toMillis(),
+        PROBE_INTERVAL.toMillis(),
+        TimeUnit.MILLISECONDS);
+    Duration initialDelayDetect =
+        Duration.ofMillis(ThreadLocalRandom.current().nextLong(PROBE_INTERVAL.toMillis()));
+    executor.scheduleAtFixedRate(
+        this::detectAndRemoveOutlierEntries,
+        initialDelayDetect.toMillis(),
+        PROBE_INTERVAL.toMillis(),
```

Comment:
I think we can schedule this task every MIN_EVICTION_INTERVAL millis, and we won't need the check in detectAndRemoveOutlierEntries?